File: src/main/java/me/zeroeightsix/kami/module/modules/misc/MidClickFriends.java
package me.zeroeightsix.kami.module.modules.misc;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.command.commands.FriendCommand;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.util.Friends;
import net.minecraft.client.Minecraft;
import net.minecraft.client.entity.EntityOtherPlayerMP;
import net.minecraft.entity.Entity;
import net.minecraft.util.math.RayTraceResult;
import net.minecraftforge.fml.common.gameevent.InputEvent;
import org.lwjgl.input.Mouse;

import static me.zeroeightsix.kami.util.MessageSendHelper.sendChatMessage;

/**
 * @author Indrit
 * Updated by Indrit on 02/03/20
 */
@Module.Info(
        name = "MidClickFriends",
        category = Module.Category.MISC,
        description = "Middle click players to friend or unfriend them",
        showOnArray = Module.ShowOnArray.OFF
)
public class MidClickFriends extends Module {
    private int delay = 0;

    @Override
    public void onUpdate() {
        if (delay > 0) {
            delay--;
        }
    }

    @EventHandler
    public Listener<InputEvent.MouseInputEvent> mouseListener = new Listener<>(event -> {
        if (delay == 0) {
            if (Mouse.getEventButton() == 2) { // 0 is left, 1 is right, 2 is middle
                if (Minecraft.getMinecraft().objectMouseOver.typeOfHit.equals(RayTraceResult.Type.ENTITY)) {
                    Entity lookedAtEntity = Minecraft.getMinecraft().objectMouseOver.entityHit;
                    if (!(lookedAtEntity instanceof EntityOtherPlayerMP)) {
                        return;
                    }
                    if (Friends.isFriend(lookedAtEntity.getName())) {
                        remove(lookedAtEntity.getName());
                    } else {
                        add(lookedAtEntity.getName());
                    }
                }
            }
        }
    });

    private void remove(String name) {
        delay = 20;
        Friends.Friend friend = Friends.friends.getValue().stream().filter(friend1 -> friend1.getUsername().equalsIgnoreCase(name)).findFirst().get();
        Friends.friends.getValue().remove(friend);
        sendChatMessage("&b" + friend.getUsername() + "&r has been unfriended.");
    }

    private void add(String name) {
        delay = 20;
        new Thread(() -> {
            Friends.Friend f = new FriendCommand().getFriendByName(name);
            if (f == null) {
                sendChatMessage("Failed to find UUID of " + name);
                return;
            }
            Friends.friends.getValue().add(f);
            sendChatMessage("&b" + f.getUsername() + "&r has been friended.");
        }).start();
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/misc/MidClickFriends.kt
package me.zeroeightsix.kami.module.modules.misc

import kotlinx.coroutines.launch
import me.zeroeightsix.kami.manager.managers.FriendManager
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.util.TickTimer
import me.zeroeightsix.kami.util.text.MessageSendHelper
import me.zeroeightsix.kami.util.threads.defaultScope
import net.minecraft.client.entity.EntityOtherPlayerMP
import net.minecraft.util.math.RayTraceResult
import net.minecraftforge.fml.common.gameevent.InputEvent
import org.kamiblue.event.listener.listener
import org.lwjgl.input.Mouse

object MidClickFriends : Module(
    name = "MidClickFriends",
    category = Category.MISC,
    description = "Middle click players to friend or unfriend them",
    showOnArray = false
) {
    private val timer = TickTimer()
    private var lastPlayer: EntityOtherPlayerMP? = null

    init {
        listener<InputEvent.MouseInputEvent> {
            // 0 is left, 1 is right, 2 is middle
            if (Mouse.getEventButton() != 2 || mc.objectMouseOver == null || mc.objectMouseOver.typeOfHit != RayTraceResult.Type.ENTITY) return@listener
            val player = mc.objectMouseOver.entityHit as? EntityOtherPlayerMP ?: return@listener
            if (timer.tick(5000L) || player != lastPlayer && timer.tick(500L)) {
                if (FriendManager.isFriend(player.name)) remove(player.name)
                else add(player.name)
                lastPlayer = player
            }
        }
    }

    private fun remove(name: String) {
        if (FriendManager.removeFriend(name)) {
            MessageSendHelper.sendChatMessage("&b$name&r has been unfriended.")
        }
    }

    private fun add(name: String) {
        defaultScope.launch {
            if (FriendManager.addFriend(name)) MessageSendHelper.sendChatMessage("Failed to find UUID of $name")
            else MessageSendHelper.sendChatMessage("&b$name&r has been friended.")
        }
    }
}